hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/FeeManager.java

```diff
@@ -135,9 +135,11 @@ public ResponseCodeEnum update(@NonNull final Bytes bytes) {
             currentSchedule = FeeSchedule.DEFAULT;
         }
 
-        // Populate the map of HederaFunctionality -> FeeData for the current schedule
-        this.currentFeeDataMap = new HashMap<>();
-        populateFeeDataMap(currentFeeDataMap, currentSchedule.transactionFeeSchedule());
+        // Populate the map of HederaFunctionality -> FeeData for the current schedule, but avoid mutating
+        // the active one in-place as other threads may be using it for ingest/query fee calculations
+        final var newCurrentFeeDataMap = new HashMap<Entry, FeeData>();
+        populateFeeDataMap(newCurrentFeeDataMap, currentSchedule.transactionFeeSchedule());
+        this.currentFeeDataMap = newCurrentFeeDataMap;
```

Comment:
This is interesting. It looks like the issue is that FeeManager needs to be threadsafe, but is not. I don't think this is sufficient to actually ensure thread safety, because the fields themselves are not either `AtomicReference` or `transient`, or guarded by synchronization. It is possible two CPUs could see different references to the hashmaps for an unspecified amount of time (some memory barrier must be crossed forcing the CPUs to flush their local caches and re-read the shared state).

Reply:
True, because fee schedules updates are _extremely_ infrequent, I didn't want to add any synchronization for the general case of safe concurrent reads from ingest/query threads and the handle thread.

This change just ensures that _if_ handle is changing the fee schedules while an ingest/query thread reading `FeeData`, the ingest/query thread will at least see a consistent set of schedules instead of discovering an empty `HashMap` with completely missing `FeeData` and logging a scary warning.